### PR TITLE
Missileguidance -- rewrite to use real-world guidance principles, general refactor

### PR DESCRIPTION
**Pull request by TheCandianVendingMachine**
Updated PR, based on #8215 
--------
## Overview

**When merged this pull request will:**
- Add real navigation profiles to all missiles. Uses something called "proportional navigation" - a control theory that suggests that as long as your line of sight angle isn't changing you are guaranteed to be on a collision course ([See this video for a demonstration within World in Conflict](https://youtu.be/Osb7anMm1AY)). This automatically leads targets and allows for insane accuracy. [This picture shows a javelin attacking an incorrectly scripted target (`getPos` instead of `centerOfMass`) which was moving at 100kmh at a distance of 2km at a firing angle of 90 degrees](https://i.imgur.com/Q7GmeIc.jpg)

Navigation types:
- Direct Line of Sight
- Proportional Navigation
- Augmented Proportional Navigation (takes into account target acceleration)
- Zero-Effort Miss (separate form of proportional navigation that allows for a more efficient intercept)

Changes flight model of weapons to utilise in-engine calculations of velocity. This avoids issues where the missile will lose all of its energy and hang in the air. This also allows for "bang-bang" guidance weapons, such as the GBU-12.

## Merge Strategy

In the linked project, all components labelled as “Necessary” must be merged at the same time; otherwise there will be broken components. “Supplementary” items are features which add new missiles to fully cover all vanilla systems, but otherwise are not breaking. “Nice to have” are features which were developed off of this branch, but are basically their own independent PR.

## Misc

To-Do
- [x] Localisation
- [x] Documentation
- [X] Add NLAW navigation type
- [X] Add SACLOS/Beam guided navigation type
- [x] Add Hellfire mid-course guidance logic
- [x] Tweak Javelin mid-course guidance logic
- [x] Fix RPT spam through IR seeker when close to target
- [X] Create UX for selecting GPS attack profile/mission (Between PP and TOO, and between PP missions)
- [X] Add ground based missiles (see todo.txt)
- [x] Multiplayer testing
- [x] Multiplayer tweaking
- [x] Convert all tabs to spaces (text editor reset settings apparently)
- [X] Add videos of this system in action

Videos:
https://youtu.be/5MrPyqcZc0I
https://youtu.be/WF9K1GitV00
https://youtu.be/xXSJLwrJW1A
https://youtu.be/E9O9M5OcxdU
https://youtu.be/_v6BdAw-MA0

### CODE REVIEW

File: addons/missileguidance/functions/fnc_navigationType_zeroEffortMiss.sqf [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+};
+
+private _zeroEffortMiss = _vectorToTarget vectorAdd (_closingVelocity vectorMultiply _timeToGo);
+private _zeroEffortMissProjectiled = _attackProfileDirection vectorMultiply (_zeroEffortMiss vectorDotProduct _attackProfileDirection);
+private _zeroEffortMissNormal = _zeroEffortMiss vectorDiff _zeroEffortMissProjectiled;
+
+private _commandedAcceleration = _zeroEffortMissNormal vectorMultiply (_navigationGain / (_timeToGo * _timeToGo));
+
+if (accTime > 0) then {
+    _navigationParams set [0, [_lineOfSight]];
```
Comment (PabstMirror) [inline addons/missileguidance/functions/fnc_navigationType_zeroEffortMiss.sqf:39]: Not sure what `_lineOfSight`

Comment (TheCandianVendingMachine) [reply addons/missileguidance/functions/fnc_navigationType_zeroEffortMiss.sqf:39]: Appears to be leftovers from a previous iteration, the line of sight isnt used in ZEM. Can be removed

File: addons/missileguidance/functions/fnc_onFired.sqf [hunk 2]
```diff
@@ -19,9 +19,6 @@
 
 params ["_shooter","_weapon","","_mode","_ammo","","_projectile"];
 
-// Bail on not missile
-if (!(_ammo isKindOf "MissileBase")) exitWith {};
```
[the remark this answers is not in the document]
Comment (TheCandianVendingMachine) [reply addons/missileguidance/functions/fnc_onFired.sqf:23]: Yeah, pretty much. There is also not really any reason to disallow other projectile types anyway, although I get it is a small optimization to each bullet fired

File: addons/missileguidance/XEH_postInit.sqf [hunk 3]
```diff
@@ -2,12 +2,13 @@
...
     false
-},
-{
+}, {
     false
 },
 [15, [false, true, false]], false] call CBA_fnc_addKeybind;  //Ctrl+Tab Key
+
+GVAR(dev_fnc_projectileCamera) = compile preprocessFileLineNumbers QPATHTOF(dev\projectileCamera.sqf);
+
```
[the remark this answers is not in the document]
Comment (PabstMirror) [reply addons/missileguidance/XEH_postInit.sqf:14]: right now there is a gvar you can set for debugging info
if we're going to keep that then we should just move this into a standard `prep(`

**Review by PabstMirror — APPROVED** [review]

**Pull request merged**